2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c

```diff
@@ -0,0 +1,193 @@
+  }
+
+  void addFrame(Frame toAdd){
+      toAdd.previous = root.previous;
+      toAdd.next = &root;
+      root.previous = &toAdd;
+  }
+
+  char * frameToString(Frame frame){
+    char * string = malloc(50);
+    sprintf(string,"%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d",
+      frame.analog_main[0]+127,
+      frame.analog_main[1]+127,
+      frame.analog_main[2]+127,
+      frame.analog_main[3]+127,
+      frame.digital_main[0],
+      frame.digital_main[1],
+      frame.digital_main[2],
+      frame.digital_main[3],
+      frame.digital_main[4],
+      frame.digital_main[5],
+      frame.digital_main[6],
+      frame.digital_main[7],
+      frame.digital_main[8],
+      frame.digital_main[9],
+      frame.digital_main[10],
+      frame.digital_main[11],
+      frame.analog_partner[0]+127,
+      frame.analog_partner[1]+127,
+      frame.analog_partner[2]+127,
+      frame.analog_partner[3]+127,
+      frame.digital_partner[0],
+      frame.digital_partner[1],
+      frame.digital_partner[2],
+      frame.digital_partner[3],
+      frame.digital_partner[4],
+      frame.digital_partner[5],
+      frame.digital_partner[6],
+      frame.digital_partner[7],
+      frame.digital_partner[8],
+      frame.digital_partner[9],
+      frame.digital_partner[10],
+      frame.digital_partner[11]
+    );
+    // TODO: code above is really ugly and inefficient lol
+    printf("%s\n",string);
+    return string;
+  }
+
+  Frame stringToFrame(char* string){
+    Frame frame = NULLFRAME;
+    char* analog_main = substring(string,1,12);
+    char* digital_main = substring(string,13,12);
+    char* analog_partner = substring(string,25,12);
+    char* digital_partner = substring(string,37,12);
+
+    // printf("%d\t%d\t%d\t%d\t",frame.analog_main[0],frame.analog_main[1],frame.analog_main[2],frame.analog_main[3]);
+
+    int x;
+    for(x=0;x<4;x++){
+      frame.analog_main[x] = atoi(substring(analog_main,(x*3)+1,3))-127;
+      frame.analog_partner[x] = atoi(substring(analog_partner,(x*3)+1,3)-127);
+    }
+    for(x=0;x<12;x++){
+      frame.digital_main[x] = atoi(substring(digital_main,x+1,1));
+      frame.digital_partner[x] = atoi(substring(digital_partner,x+1,1));
+    }
+
+     printf("%d\t%d\t%d\t%d\n",frame.analog_main[0],frame.analog_main[1],frame.analog_main[2],frame.analog_main[3]);
+     printf("%d",frame.digital_main[0]);
+    // printf("%d\t%d\t%d\t%d\t",NULLFRAME.analog_main[0],NULLFRAME.analog_main[1],NULLFRAME.analog_main[2],NULLFRAME.analog_main[3]);
+
+    return frame;
+
+  }
+
+  char* substring(char* str,int start,int length){
+    /*
+    *char *substring = malloc(length+1);
+    strncpy(substring, str+1, length);
+    return substring;
+    */ // Code from Mikel, returns 712 instead of 127
+
+    char *substring = malloc(length+1);
+    int subPos;
+
+    if(substring == NULL){
+       printf("str in substring method is NULL.\n");
+       exit(1);
+    }
+
+    for(subPos = 0 ; subPos < length ; subPos++){
```

Comment:
not a huge fan of this implementation, I think the strncpy version on #electrical will be cleaner and faster (worth it to benchmark speeds or nah?)